cmd/patchVolume.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright © 2019 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cmd
+
+import (
+	"fmt"
+
+	api "github.com/libopenstorage/openstorage-sdk-clients/sdk/golang"
+	"github.com/portworx/px/pkg/portworx"
+	"github.com/portworx/px/pkg/util"
+	"github.com/spf13/cobra"
+)
+
+type volumeUpdateOpts struct {
+	req        *api.SdkVolumeUpdateRequest
+	halevel    int64
+	replicaSet []string
+	size       uint64
+	shared     bool
+	sticky     bool
+}
+
+var (
+	updateReq      *volumeUpdateOpts
+	patchVolumeCmd *cobra.Command
+)
+
+// updateVolumeCmd represents the updateVolume command
+var _ = RegisterCommandVar(func() {
+	updateReq = &volumeUpdateOpts{
+		req: &api.SdkVolumeUpdateRequest{},
+	}
+
+	patchVolumeCmd = &cobra.Command{
+		Use:   "volume [NAME]",
+		Short: "Update field(s) of a portworx volume",
+		Example: `$ px patch  volume test1 --halevel 3
+	This set halevel of volume to 3.`,
+
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return fmt.Errorf("Must supply a name for the volume")
+			}
+			return nil
+		},
+		RunE: updateVolume,
+	}
+})
+
+var _ = RegisterCommandInit(func() {
+	patchCmd.AddCommand(patchVolumeCmd)
+
+	patchVolumeCmd.Flags().Int64Var(&updateReq.halevel, "halevel", 0, "New replication factor (Valid Range: [1, 3]) (default 1)")
+	patchVolumeCmd.Flags().StringSliceVarP(&updateReq.replicaSet, "nodes", "n", []string{}, "")
+	patchVolumeCmd.Flags().Uint64VarP(&updateReq.size, "size", "s", 0, "New size for the volume (GiB) (default 1)")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "shared", "r", false, "Shared is true if this volume can be remotely accessed [default false].")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "sticky", "t", false, "set sticky setting [default false].")
+	patchVolumeCmd.Flags().SortFlags = false
+})
+
+func updateVolume(cmd *cobra.Command, args []string) error {
+	ctx, conn, err := PxConnectDefault()
+	ok := false
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	// fetch the volume name from args
+	updateReq.req.VolumeId = args[0]
+
+	// By default blank replicaset needs to be passed due to know volume driver issuw (pwx-9500)
+	// If user provides one it will be overriden
+	updateReq.req.Spec = &api.VolumeSpecUpdate{
```

Comment:
There should be an `if` before this line, if there `len (updateReq.replicaSet) != 0`

Reply:
Reason for not having an if() check is "replicaSet" always needs to be passed or leads to PWX-9500 issue.  So either it's empty or user-provided, by default it's empty.

Reply:
As discussed passing replicaSet only if halevel is set.